src/axom/inlet/examples/nested_structs.cpp

```diff
@@ -0,0 +1,276 @@
+    table.addString("path");
+    // A string is used to represent the enumeration
+    table.addString("units");
+    table.addInt("start_dimensions");
+    // addGenericArray is used to represent std::vector<T> where
+    // T is any non-primitive type
+    auto& ops_table = table.addGenericArray("operators");
+    Operator::defineSchema(ops_table);
+  }
+};
+
+template <>
+struct FromInlet<Geometry>
+{
+  Geometry operator()(const inlet::Table& base)
+  {
+    Geometry result;
+    result.format = base["format"];
+    result.path = base["path"];
+    result.start_dim = base["start_dimensions"];
+
+    // FIXME: Remove when PR #429 is merged
+    auto ops = base["operators"].get<std::unordered_map<int, Operator>>();
+    for(const auto& ele : ops)
+    {
+      result.operators.push_back(ele.second);
+    }
+    return result;
+  }
+};
+
+std::ostream& operator<<(std::ostream& os, const Geometry& geom)
+{
+  os << fmt::format("Geometry in format: '{0}'\n", geom.format);
+  os << fmt::format("  with path: '{0}'\n", geom.path);
+  for(const auto& op : geom.operators)
+  {
+    os << op;
+  }
+  return os;
+}
+
+struct Shape
+{
+  std::string name;
+  enum class Material
+  {
+    Steel,
+    Wood,
+    Plastic
+  } material;
+  Geometry geom;
+  static void defineSchema(inlet::Table& table)
+  {
+    table.addString("name");
+    table.addString("material");
+    // addStruct is used for a single instance of a user-defined type
+    auto& geom_table = table.addStruct("geometry");
+    Geometry::defineSchema(geom_table);
+  }
+};
+
+std::ostream& operator<<(std::ostream& os, const Shape& shape)
+{
+  os << fmt::format("Shape: '{0}'\n", shape.name);
+  os << shape.geom;
+  return os;
+}
+
+template <>
+struct FromInlet<Shape>
+{
+  Shape operator()(const inlet::Table& base)
+  {
+    Shape result;
+    result.name = base["name"];
+    std::string material = base["material"];
+    if(material == "steel")
+    {
+      result.material = Shape::Material::Steel;
+    }
+    result.geom = base["geometry"].get<Geometry>();
+    return result;
+  }
+};
+
+const std::string input = R"(
+dimensions: 2
+shapes:
+  - name: wheel
+    material: steel
+    geometry:
+      format: test_format
+      path: path/to/file.format
+      units: cm
+      start_dimensions: 3
+      operators:
+        - translate: [10, 20, 30]
+        - rotate: 45
+          axis: [1, 2, 3]
+          center: [4, 5, 6]
+        - slice:
+            x: 10
+        - translate: [5, 6]
+)";
+
+int main()
+{
+  // Inlet requires a SLIC logger to be initialized to output runtime information
+  axom::slic::SimpleLogger logger;
+
+  axom::sidre::DataStore ds;
+  auto reader = std::unique_ptr<inlet::YAMLReader>(new inlet::YAMLReader());
+  reader->parseString(input);
+  inlet::Inlet inlet(std::move(reader), ds.getRoot());
+
+  auto& shapes_table = inlet.addGenericArray("shapes");
+  Shape::defineSchema(shapes_table);
+
+  auto shapes = inlet["shapes"].get<std::unordered_map<int, Shape>>();
```

Comment:
It would also be interesting to see what the generated sphinx docs would look like for this nested structs example. Would it be easy to add a "generate docs" option?

Reply:
Yes, although this appears to have exposed some inadequacies in the SphinxDocWriter logic.  I have logged this as #436